### ISSUE TITLE
[vscode]: TypeError: cannot serialize '_io.TextIOWrapper' (pydevd_bundle.pydevd_io.IORedirector)

### ISSUE THREAD

**Issue by mike-pt** [microsoft/ptvsd#1607]
## Environment data

- PTVSD version: (not sure, using: ms-python.python-2019.6.24221 on vscode)
- OS and version: macOS Mojave and Win 10
- Python version (& distribution if applicable, e.g. Anaconda): 3.7.4 in a venv
- Using VS Code or Visual Studio: VSCODE

## Actual behavior

When I try to debug a python app with multiprocessing (pathos) and using spawn, I'm seeing this when we use pool.map:
`TypeError: cannot serialize '_io.TextIOWrapper' obj`

I added a try except where this happens to figure which object actually causes it (since this runs fine from terminal) and I get this:

`TypeError: cannot serialize '_io.TextIOWrapper' $<_pydevd_bundle.pydevd_io.IORedirector object at 0x1073aa3d0>`

## Expected  behavior

Code should run fine as it does in terminal

## Steps to reproduce:
1. Unsure we pass `pool.map(self.case_list_exec, node_args)`  (the case_list_exec function executes a list of test cases)

**Comment by int19h** [comment]
This sounds similar to https://github.com/microsoft/ptvsd/issues/1283#issuecomment-501127755 - basically, you need to make sure that whatever data you're sharing between your processes (even inadvertently) is serializable when using "spawn".

**Comment by mike-pt** [comment]
Yeah but this is not data we are passing, it's something the debugger adds.

**Comment by int19h** [comment]
This looks like the debugger replacement object for `sys.stdout` and `sys.stderr` (that is used to redirect output into the IDE). It shouldn't be passed between processes by itself, so it's likely due to the structure of the code that ends up inadvertently placing it in the graph of objects that cross the boundary - it's unfortunately rather easy to do this with `multiprocessing`.

Can you show the complete call stack of the error, and ideally, the code on which it happens?

**Comment by mike-pt** [comment]
Hum I did original though of the logger as the problem, however this works outside of vscode without any changes.

We also attempted to comment out the logger parts so it wouldn't be passed, but perhaps we missed something, but that would be the only thing that makes sense right? 

We share logging and I didn't work on that code but it should be locking and not passing that between processes (perhaps I need to look deeper into that though).


In any case here is the current stack trace after I changed the inner python module with a try except so it would print the object that can't be serialized (which feels like its something python should do).

Anyway here is the stack trace now (I can share the full code since this is private, but I think its fine to show the stack

Original:
```
cannot serialize '_io.TextIOWrapper' object
Traceback (most recent call last):
  File "/Users/user/gitstuff/qaenv/tools/testTools/testTrigger.py", line 52, in testing
    exitcode = execution.execution()
  File "/Users/user/gitstuff/qaenv/tools/executionTools/testExecution.py", line 55, in execution
    self.config.resultDict = self.platformExecution()
  File "/Users/user/gitstuff/qaenv/tools/executionTools/testExecution.py", line 167, in platformExecution
    driverResults = pool.map(self.__caseListExecution, nodeArgs)
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/pool.py", line 657, in get
    raise self._value
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/pool.py", line 431, in _handle_tasks
    put(task)
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/connection.py", line 209, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/reduction.py", line 54, in dumps
    cls(buf, protocol).dump(obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 437, in dump
    self.save(obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 793, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 778, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 778, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 1086, in save_instancemethod0
    pickler.save_reduce(MethodType, (obj.__func__, obj.__self__), obj=obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 642, in save_reduce
    save(args)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 778, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 553, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 666, in save_reduce
    save(state)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 910, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 863, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 889, in _batch_setitems
    save(v)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 553, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 666, in save_reduce
    save(state)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 910, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 863, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 889, in _batch_setitems
    save(v)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 823, in save_list
    self._batch_appends(obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 847, in _batch_appends
    save(x)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 910, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 863, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 889, in _batch_setitems
    save(v)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 823, in save_list
    self._batch_appends(obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 847, in _batch_appends
    save(x)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 553, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 666, in save_reduce
    save(state)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 910, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 863, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 889, in _batch_setitems
    save(v)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 525, in save
    rv = reduce(self.proto)
TypeError: cannot serialize '_io.TextIOWrapper' object
```

And this is after I added the try execpt on the pickle.py code so it would print the object that can't be serialized:
```
TypeError: cannot serialize '_io.TextIOWrapper' $<_pydevd_bundle.pydevd_io.IORedirector object at 0x10bd8be50>
<built-in method __reduce_ex__ of IORedirector object at 0x10bd8be50> must return string or tuple
Traceback (most recent call last):
  File "/Users/user/gitstuff/qaenv/tools/testTools/testTrigger.py", line 52, in testing
    exitcode = execution.execution()
  File "/Users/user/gitstuff/qaenv/tools/executionTools/testExecution.py", line 55, in execution
    self.config.resultDict = self.platformExecution()
  File "/Users/user/gitstuff/qaenv/tools/executionTools/testExecution.py", line 167, in platformExecution
    driverResults = pool.map(self.__caseListExecution, nodeArgs)
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/pool.py", line 657, in get
    raise self._value
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/pool.py", line 431, in _handle_tasks
    put(task)
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/connection.py", line 209, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/multiprocess/reduction.py", line 54, in dumps
    cls(buf, protocol).dump(obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 437, in dump
    self.save(obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 794, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 779, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 779, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 1086, in save_instancemethod0
    pickler.save_reduce(MethodType, (obj.__func__, obj.__self__), obj=obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 643, in save_reduce
    save(args)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 779, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 554, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 667, in save_reduce
    save(state)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 910, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 864, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 890, in _batch_setitems
    save(v)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 554, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 667, in save_reduce
    save(state)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 910, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 864, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 890, in _batch_setitems
    save(v)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 824, in save_list
    self._batch_appends(obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 848, in _batch_appends
    save(x)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 910, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 864, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 890, in _batch_setitems
    save(v)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 824, in save_list
    self._batch_appends(obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 848, in _batch_appends
    save(x)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 554, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 667, in save_reduce
    save(state)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/user/.virtualenvs/qaenv/lib/python3.7/site-packages/dill/_dill.py", line 910, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 864, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 890, in _batch_setitems
    save(v)
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pickle.py", line 545, in save
    raise PicklingError("%s must return string or tuple" % reduce)
_pickle.PicklingError: <built-in method __reduce_ex__ of IORedirector object at 0x10bd8be50> must return string or tuple
```

**Comment by mike-pt** [comment]
This was also code that I didn't write so I'm trying to dig into it myself too, but it looks like we do try to handle logging so it works in multiprocessing (and again works fine on terminal)

**Comment by mike-pt** [comment]
Seems like the code is pretty similar to this: https://mattgathu.github.io/multiprocessing-logging-in-python/ for the log handler, we use a queue, so I'm back to not really understanding why this is failing

**Comment by int19h** [comment]
Judging by the stack, it looks like it's trying to pickle the dict of a module (note `save_module_dict`) - so that object is probably referenced by some global variable in a module.

However, it's also interesting that the call stack is mostly stock `pickle`, but for this particular frame, it's `dill`. I've looked that up, and it appears to be some kind of custom serializer that `pathos` introduces? This might be why it's not observable in regular multiprocessing apps - it looks like it might be trying to serialize more things than regular `pickle` does, and so your graph ends up larger. Their docs particularly mention the ability to serialize function closures.

The interesting parts in the stack are these frames:
- `save_instancemethod0`
- `save_module_dict`
- `save_module_dict`
- `save_module_dict`

It looks like your `node_args` contain some object, for which `dill` needed to serialize an instance method to pass it across to the other node - that's the red flag for these types of scenarios. You generally don't want to be serializing code, and especially not closure-like objects (in case of instance methods, they're effectively closures over `self`). It's likely that the `self` of this method was something that ends up dragging this entire object graph, which includes modules etc (likely the ones that the object to which it belongs uses). And one of those modules has a global variable that references `sys.stdout` or `sys.stderr` - probably the logger module, since it's a very common pattern for them.

The article that you link to about logging describes how to do it with stock `multiprocessing`, which would only use `pickle`, so I think that's at least part of the reason why they did not run into issues like this. You can still hit stuff like that with `pickle`, but it's much more obvious as to what it is about, because it will fail as soon as it sees a closure, rather than trying to serialize the data that closure references.

Why it works outside of debugger is an interesting question. I suspect that `dill` has some special-casing code that specifically handles `sys.stdout` etc - they might have run into this issue previously, and this is the most obvious fix, since you don't really need to serialize those objects at all; all processes will share stdio streams anyway. But if they do it by checking the object type, then it would fail in this case, because debugger stdio wrappers have a different type, of course. 

Either way, you need to find that object that needs to serialize the instance method - that is the culprit. Dill appears to have a [special module](https://dill.readthedocs.io/en/latest/dill.html#module-dill.detect) for diagnosing these types of issues; you can use it on your `node_args`, and see what you get out of `badobjects` - this should at least tell you _which_ instance method it is, and then it should be easy to locate the place where you capture it in a variable in your code.

**Comment by mike-pt** [comment]
@int19h  thanks for the detailed analyses and yes this is using pathos which uses dill for most things (unless when it can't) which would be the cases for logger.

We though this was handled correctly case it worked on the terminal (at least linux/osx) and that was mostly based on that article, but like you said this is different with pathos.

We will probably move back to mutiprocessing but for now we tracked it down and the issue was that with this approach we were basically creating new logs and then appending to the existing one, and while doing so also reseting the handlers to the original ones for the other (none shared logs) and this makes it none serialize.

Python 3 actually as a different handler for queue that solves this, but It turns out we also don't need to do what we were doing anyway:

So we removed this: 
```
        if not logConfig['log'].handlers:
          logConfig['log'].handlers = logConfig['logHandlers']

```

And where we were doing:
```
      logConfig['log'] = logger.changeLogMultiProcessState(config, logConfig['log'], logConfig['logName'], logConfig['logQueue'])
```

We now only need
```
      logConfig['log'] = logger.changeLogMultiProcessState(config, logConfig['log'], queue=logConfig['logQueue'])
```
We also remove the portion of the code where we were saving the original file handlers and then re-adding them (turn out we don't need that anyway) and this way we avoid passing none seriazable objects.


I suppose its complicated  to  explain without the full code, but the tl;dr is what you've already said. We could also opt to use queueHandler since this is python3.


Anyway I'm closing this as its not a debbuger issue, and I hope it at least helps other folks :)
